src/workerd/server/server.c++

```diff
@@ -1380,12 +1387,86 @@ public:
       return kj::heap<ActorChannelImpl>(*this, kj::mv(id));
     }
 
+    // ActorContainer mostly serves as a wrapper around Worker::Actor.
+    // We use it to associate a HibernationManager with the Worker::Actor, since the
+    // Worker::Actor can be destroyed during periods of prolonged inactivity.
+    //
+    // We use a RequestTracker to track strong references to this ActorContainer's Worker::Actor.
+    // Once there are no Worker::Actor's left (excluding our own), `inactive()` is triggered and we
+    // initiate the eviction of the Durable Object. If no requests arrive in the next 10 seconds,
+    // the DO is evicted, otherwise we cancel the eviction task.
+    class ActorContainer final: public RequestTracker::Hooks {
+    public:
+      ActorContainer(kj::String key, ActorNamespace& parent, kj::Timer& timer)
+          : tracker(kj::refcounted<RequestTracker>(*this)),
+            key(kj::mv(key)),
+            parent(parent),
+            timer(timer) {}
+
+      ~ActorContainer() {
+        // We need to make sure we're removed from the actors map.
+        parent.actors.erase(key);
+      }
+
+      void active() override {
+        // We're handling a new request, cancel the eviction promise.
+        shutdownTask = nullptr;
+      }
+
+      void inactive() override {
+        KJ_IF_SOME(a, actor) {
+          KJ_IF_MAYBE(m, a->getHibernationManager()) {
+            // The hibernation manager needs to survive actor eviction and be passed to the actor
+            // constructor next time we create it.
+            manager = kj::addRef(*static_cast<HibernationManagerImpl*>(m));
+          }
+        }
+        // After 10 seconds of inactivity, we destroy the Worker::Actor and hibernate any active
+        // JS WebSockets.
+        shutdownTask = timer.afterDelay(10 * kj::SECONDS).then([&]() -> kj::Promise<void> {
```

Comment:
Can this be reworked to use coroutine syntax instead?

Reply:
I can give it another go in the morning but it's not looking too promising.
```
error: this function cannot be a coroutine: 'std::coroutine_traits<void, workerd::server::Server::WorkerService::ActorNamespace::ActorContainer &>' has no member named 'promise_type'
      void inactive() override {
```

I think `inactive` needs to return a promise, but we're overriding this method from an abstract class that's extended elsewhere too. If so I'd rather defer that and tackle all implementations of `inactive()`  at once.